python_interface/design/03_pyHorace_description.md

```diff
@@ -0,0 +1,269 @@
+because it causes a second Python interpreter to started within the same process as the Matlab process 
+(which is actually the same process as the parent *Python* process because we are loading Matlab "in-process"). 
+Having two Python interpreters in the same process invariably causes a memory error and hard crash.
+There is no restriction on including `py.*` code in a compiled Matlab application 
+because they can be safely handled if the Matlab application is called directly from the commandline and not loaded as a Python module.
+A possible work-around is create the second Python interpreter (launched by `py.*`) in a new process ("out-of-process" execution). 
+However testing shows that this imposes huge performance penalties (using Brille becomes an order of magnitude slower).
+Furthermore for `pyHorace` the second interpreter is not connected with the user's interpreter
+so cannot know about any fitting/modelling functions they have defined.
+
+For these reasons, `pyHorace` opted to use a `mex` [file](https://github.com/mducle/hugo/blob/master/src/call_python.cpp) which is
+called by Matlab instead of the Python function input. 
+This mex file is linked to the CPython API so it can access the parent Python interpreter which is loaded into the same process.
+(Thus this method will not work with an out-of-process Python interpreter).
+
+First when the `DataTypes` class encounters a Python callable (a function or class which can be called), instead of passing this to Matlab,
+it stores the function reference in a [module-global dictionary](https://github.com/mducle/hugo/blob/master/pyHorace/FunctionWrapper.py#L21),
+and then creates a Matlab [`pythonFunctionWrapper`](https://github.com/mducle/hugo/blob/master/src/pythonFunctionWrapper.m) object
+which stores the (string) key to the Python function.
+Whenever `call_method` encounters a `pythonFunctionWrapper` object,
+it [wraps](https://github.com/mducle/hugo/blob/master/src/check_wrapped_function.m) this in a call to the mex `call_python` function with the correct key.
+Thus it appears to the user that Matlab is calling a Python function transparently (but actually happens is that Matlab calls `call_python`).
+`call_python` then [retrieves](https://github.com/mducle/hugo/blob/master/src/call_python.cpp#L390) the Python function reference from the module-global `dict` 
+and [calls](https://github.com/mducle/hugo/blob/master/src/call_python.cpp#L413) that function with the arguments given to it.
+
+The majority of the code in `call_python` is for data conversion between Matlab and Python.
+This is because `mex` functions may only take and return *Matlab* data types (Python objects are not allowed).
+Similarly the CPython `PyObject_Call` function requires that inputs to the called functions are all of the `PyObject` type.
+Thus `call_python` converts Matlab types to their Python equivalent and vice versa as per this table:
```

Comment:
I'm unclear how new-style classes are handled here.

Reply:
Matlab classes are currently not handled... It should certainly be possible to handle them (wrapped in a `MatlabProxyObject` Python wrapper) but this needs to be implemented (mainly because I've not needed to do this for any of the current functionality).